Allow hiding of reblogs from followed users

aschmitz opened the pull request:
This adds a new entry to the account menu to allow users to hide future reblogs from a user (and then if they've done that, to show future reblogs instead).

This does not remove or add historical reblogs from/to the user's timeline; it only affects new statuses.

The API for this operates by sending a "reblogs" key to the follow endpoint. If this is sent when starting a new follow, it will be respected from the beginning of the follow relationship (even if the follow request must be approved by the followee). If this is sent when a follow relationship already exists, it will simply update the existing follow relationship. As with the notification muting, this will now return an object (`{reblogs: [true|false]}`) or `false` for each follow relationship when requesting relationship information for an account. This should cause few issues due to an object being truthy in many languages, but some modifications may need to be made in pickier languages.

Database changes: adds a `show_reblogs` column (default `true`, non-nullable) to the `follows` and `follow_requests` tables. Because these are non-nullable, we use the existing MigrationHelpers to perform this change without locking those tables, although the tables are likely to be small anyway.

Tests included.

See also <https://github.com/glitch-soc/mastodon/pull/212>.

aschmitz commented:
(I think this closes #648, and replaces #3271.)

aschmitz commented on app/javascript/mastodon/reducers/accounts_counters.js:
```diff
@@ -126,6 +126,7 @@ export default function accountsCounters(state = initialState, action) {
   case STATUS_FETCH_SUCCESS:
     return normalizeAccountFromStatus(state, action.status);
   case ACCOUNT_FOLLOW_SUCCESS:
+    if (action.alreadyFollowing) { return state; }
```
eslint hates this line. I can make it ternary if you prefer, but that seemed *less* clear, not more.

Gargron replied:
You can make it a function call instead

Gargron commented:
>(I think this closes #648, and replaces #3271.)

To have it automatically do this, write "fix #648, resolve #3271" in the body of your PR (or any commit message)

Gargron commented on app/lib/feed_manager.rb:
```diff
@@ -165,7 +165,9 @@ def filter_from_home?(status, receiver_id)
       return should_filter
     elsif status.reblog?                                                                                                 # Filter out a reblog
-      should_filter   = Block.where(account_id: status.reblog.account_id, target_account_id: receiver_id).exists?        # or if the author of the reblogged status is blocking me
+      src_id = status.account_id
```
Why is this a variable?

Gargron commented on db/migrate/20171028221157_add_reblogs_to_follows.rb:
```diff
@@ -0,0 +1,21 @@
+class AddReblogsToFollows < ActiveRecord::Migration[5.1]
+  include Mastodon::MigrationHelpers
+
+  safety_assured do
```
This wrapping should not be necessary

aschmitz replied:
Hm, right you are. I thought strong_migrations complained otherwise, guess not(? or at least not when I just tried it).

Gargron commented on app/controllers/api/v1/accounts_controller.rb:
```diff
@@ -13,9 +13,11 @@ def show
 
   def follow
-    FollowService.new.call(current_user.account, @account.acct)
+    reblogs_arg = { reblogs: params[:reblogs] }
```
This could be a private method

aschmitz replied:
In theory, yes? I already think it's not great because we're using it as an options argument as well as a return value (mostly for convenience and line size, because it's the same in both cases for the moment). I don't think it really has any purpose being a method, but I can make it one if you want.

Gargron commented on app/javascript/mastodon/reducers/accounts_counters.js:
```diff
@@ -126,6 +126,7 @@ export default function accountsCounters(state = initialState, action) {
   case STATUS_FETCH_SUCCESS:
     return normalizeAccountFromStatus(state, action.status);
   case ACCOUNT_FOLLOW_SUCCESS:
+    if (action.alreadyFollowing) { return state; }
```
If blocks within a switch are kind of confusing. A ternary probably *is* better, and if it's less readable I think you can multiline it.

Gargron commented on app/models/concerns/account_interactions.rb:
```diff
@@ -66,8 +74,12 @@ def follow_mapping(query, field)
-  def follow!(other_account)
-    active_relationships.find_or_create_by!(target_account: other_account)
+  def follow!(other_account, reblogs: nil)
+    reblogs = true if reblogs.nil?
```
Can this not be part of the argument's default value..?

aschmitz replied:
In fact no, apparently. (Well, it could be, but we still need to catch the nil case: it may come in from params as nil.)

Gargron commented on app/services/follow_service.rb:
```diff
@@ -6,25 +6,38 @@ class FollowService < BaseService
-  def call(source_account, uri)
+  # @param [true, false, nil] reblogs Whether or not to show reblogs, defaults to true
+  def call(source_account, uri, reblogs: nil)
+    reblogs = true if reblogs.nil?
```
Same here, doesn't this mean the default value of the argument should be true and not nil?

aschmitz replied:
Same as the other case.

Gargron commented on app/lib/feed_manager.rb:
```diff
@@ -165,7 +165,9 @@ def filter_from_home?(status, receiver_id)
       return should_filter
     elsif status.reblog?                                                                                                 # Filter out a reblog
-      should_filter   = Block.where(account_id: status.reblog.account_id, target_account_id: receiver_id).exists?        # or if the author of the reblogged status is blocking me
+      src_id = status.account_id
```
Why is this a variable? Seems like it's only used in one place.

aschmitz replied:
Mostly because if I don't do this, it pushes the comments column out even farther, which seemed undesirable, but I can do whatever you'd like with it.

aschmitz commented:
Tests pass, minus the Code Climate issue about `class_methods` in AccountInteractions being too big of a block, which I'm inclined to ignore unless you want to massively refactor that. (Passing a proc to `follow_mapping` isn't a terrible idea, but it also won't save many lines, and I think the intent of that check is something other than what it's catching here.)

I think I've addressed the review comments, but let me know if there's something you think still needs changing. Thanks!

tateisu commented:
how about faking follow status ?
can we call "hide-boost" request while follow status is faked?
